Sources/SwiftDocC/Infrastructure/External Data/OutOfProcessReferenceResolver.swift

```diff
@@ -462,23 +462,32 @@ private class LongRunningProcess: ExternalLinkResolving {
         
     func sendAndWait<Request: Codable & CustomStringConvertible, Response: Codable>(request: Request?) throws -> Response {
         if let request = request {
-            guard let requestString = String(
-                    data: try JSONEncoder().encode(request), encoding: .utf8)?.appending("\n"),
+            guard let requestString = String(data: try JSONEncoder().encode(request), encoding: .utf8)?.appending("\n"),
                   let requestData = requestString.data(using: .utf8)
             else {
                 throw OutOfProcessReferenceResolver.Error.unableToEncodeRequestToClient(requestDescription: request.description)
             }
             input.fileHandleForWriting.write(requestData)
         }
-        let response = output.fileHandleForReading.availableData
+        var response = output.fileHandleForReading.availableData
         guard !response.isEmpty else {
             throw OutOfProcessReferenceResolver.Error.processDidExit(code: Int(process.terminationStatus))
         }
         
-        do {
-            return try JSONDecoder().decode(Response.self, from: response)
-        } catch {
-            throw OutOfProcessReferenceResolver.Error.unableToDecodeResponseFromClient(response, error)
+        // It's not guaranteed that the full response will be available all at once.
+        while true {
```

Comment:
Should we add a timeout to fail the test, instead of potentially being stuck an infinite loop?

Reply:
The loop isn't an issue because it will never iterate repeatedly. 

The real issue (as mentioned in the code comment) is that reading `availableData` from an empty pipe will block until more data is written to the pipe. This means that if we have already read the complete data for the response and get `availableData` again it deadlock since the resolver will wait for more data to be written and the other process won't write more data until it has received the next request (which the resolver will never send because it's waiting).

I'll try and add some safe guards to make this less likely to happen but we'll need to update (and likely version) the communication protocol to design away the problem completely.

Reply:
I added two more checks in 0e45644b869f417d0762512254b38edfee43d3c4 to reduce the risks of deadlocks:
- The bytes read so far is some multiple of the pipe's buffer size
- The new response data chunk wasn't empty (meaning that the pipe has reached EOF)

It is still possible to deadlock, for example by sending exactly 512 bytes of invalid JSON, but it should be very uncommon in practice.